flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/operators/windowing/GroupedActiveDiscretizer.java

```diff
@@ -60,29 +62,45 @@ public void run() throws Exception {
 					groupedDiscretizers.put(key, groupDiscretizer);
 				}
 
-				groupDiscretizer.processRealElement(nextObject);
+				groupDiscretizer.processRealElement(element);
 			}
 
-		}
 
-		for (StreamDiscretizer<IN> group : groupedDiscretizers.values()) {
-			group.emitWindow();
-		}
+
 
 	}
 
 	@Override
 	public void open(org.apache.flink.configuration.Configuration parameters) throws Exception {
 		super.open(parameters);
-		centralThread = new Thread(new CentralCheck());
+		centralCheck = new CentralCheck();
+		centralThread = new Thread(centralCheck);
 		centralThread.start();
 	}
 
+	@Override
+	public void close() throws Exception {
+		super.close();
+		for (StreamDiscretizer<IN> group : groupedDiscretizers.values()) {
+			group.emitWindow();
+		}
+
+		try {
+			centralCheck.running = false;
+			centralThread.interrupt();
+			centralThread.join();
+		} catch (InterruptedException e) {
+			e.printStackTrace();
```

Comment:
Printing only the stacktrace (without the exception's message) to standard out is really not helpful at all ;)
I would wrap it in a runtime exception, or since its in a close() method, log a warning with the exception.